src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -535,6 +505,67 @@ internal async Task Open(NpgsqlTimeout timeout, bool async, CancellationToken ca
                 Break(e);
                 throw;
             }
+
+            static async Task OpenCore(NpgsqlConnector conn, SslMode sslMode, NpgsqlTimeout timeout, bool async, CancellationToken cancellationToken)
+            {
+                await conn.RawOpen(sslMode, timeout, async, cancellationToken);
+
+                var username = conn.GetUsername();
+                if (conn.Settings.Database == null)
+                    conn.Settings.Database = username;
+
+                timeout.CheckAndApply(conn);
+                conn.WriteStartupMessage(username);
+                await conn.Flush(async, cancellationToken);
+
+                var cancellationRegistration = conn.StartCancellableOperation(cancellationToken, attemptPgCancellation: false);
+                try
+                {
+                    await conn.Authenticate(username, timeout, async, cancellationToken);
+                }
+                catch (PostgresException e) when (sslMode == SslMode.Prefer && conn.IsSecure &&
+                    e.SqlState == PostgresErrorCodes.InvalidAuthorizationSpecification)
+                {
+                    cancellationRegistration.Dispose();
+                    Debug.Assert(!conn.IsBroken);
+
+                    try
+                    {
+                        conn._stream?.Dispose();
+                    }
+                    catch
+                    {
+                        // ignored
+                    }
+
+                    conn._stream = null!;
```

Comment:
We could still extract this stuff to a common method that would be called from here and from Close